llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -621,12 +730,16 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
   MapVector<StringRef, MDNode *> SCMetadata;
   SmallVector<MDNode *, 4> DefaultsMetadata;
 
+  bool IRModified = false;
+  if (Mode == HandlingMode::default_values) {
+    IRModified |= replaceSpecConstsWithDefaultValues(M);
```

Comment:
Why can't we integrate this into a regular flow of the pass? The PR duplicates some of a basic functionality, increasing maintenance cost. I though that we could simply add a new branch next to `if (Mode == HandlingMode::native)`

New path does not generate metadata with default spec constant values - I would double-check if it is indeed not used by runtime. Could be the case that it is used to handle `get_specialization_constant_value` on host.

Reply:
At the first glance, regular flow of the pass is hard to read and I didn't want to make it worse.

Reply:
Well, it might be easy to make it good.

Reply:
Yeah, I agree that that function is way too long, but perhaps it is time to outline branches of that main `if` into a separate functions to make high-level logic easier to read

Reply:
Started it here: https://github.com/intel/llvm/pull/10613

Reply:
Integrated it into the regular loop.